sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/datasketchesAggregates.scala

```diff
@@ -0,0 +1,523 @@
+            )
+          )
+        } else {
+          TypeCheckSuccess
+        }
+      case _ =>
+        DataTypeMismatch(
+          errorSubClass = "HLLSKETCH_WRONG_TYPE",
+          messageParameters = Map(
+            "functionName" -> toSQLId(prettyName),
+            "expectedSecond" -> toSQLType(IntegerType),
+            "expectedThird" -> toSQLType(StringType),
+            "actual" -> Seq(first.dataType, second.dataType, third.dataType)
+              .map(toSQLType).mkString(", ")
+          )
+        )
+    }
+  }
+
+  // From here on, these are the shared default implementations for TypedImperativeAggregate
+
+  /** Aggregate functions which utilize HllSketch instances should never return null */
+  override def nullable: Boolean = false
+
+  /**
+   * Instantiate an HllSketch instance using the lgConfigK and tgtHllType params.
+   *
+   * @return an HllSketch instance
+   */
+  override def createAggregationBuffer(): HllSketch = {
+    new HllSketch(lgConfigK, tgtHllType)
+  }
+
+  /**
+   * Evaluate the input row and update the HllSketch instance with the row's value.
+   * The update function only supports a subset of Spark SQL types, and an
+   * UnsupportedOperationException will be thrown for unsupported types.
+   *
+   * @param sketch The HllSketch instance.
+   * @param input  an input row
+   */
+  override def update(sketch: HllSketch, input: InternalRow): HllSketch = {
+    val v = first.eval(input)
+    if (v != null) {
+      first.dataType match {
+        // Update implemented for a subset of types supported by HllSketch
+        // Spark SQL doesn't have equivalent types for ByteBuffer or char[] so leave those out
+        // Leaving out support for Array types, as unique counting these aren't a common use case
+        // Leaving out support for floating point types (IE DoubleType) due to imprecision
+        // TODO: implement support for decimal/datetime/interval types
+        case IntegerType => sketch.update(v.asInstanceOf[Int])
+        case LongType => sketch.update(v.asInstanceOf[Long])
+        case StringType => sketch.update(v.asInstanceOf[UTF8String].toString)
+        case dataType => throw new UnsupportedOperationException(
+          s"A HllSketch instance cannot be updates with a Spark ${dataType.toString} type")
+      }
+    }
+    sketch
+  }
+
+  /**
+   * Merges an input HllSketch into the sketch which is acting as the aggregation buffer.
+   *
+   * @param sketch the HllSketch instance used to store the aggregation result.
+   * @param input an input HllSketch instance
+   */
+  override def merge(sketch: HllSketch, input: HllSketch): HllSketch = {
+      val union = new Union(sketch.getLgConfigK)
+      union.update(sketch)
+      union.update(input)
+      union.getResult(sketch.getTgtHllType)
+  }
+
+  /** Convert the underlying HllSketch into an updateable byte array  */
+  override def serialize(sketch: HllSketch): Array[Byte] = {
+    sketch.toCompactByteArray
+  }
```

Comment:
The comment says that we serialize to an updatable byte array, but then we serialize to a compact byte array. These need to be consistent. Also I would suggest to serialize to an updatable byte array, as serde in this case should be faster.

Reply:
Good catch - I initially had it returning an updateable byte array as I thought we'd be able to wrap/update that byte array after deserialization. I converted to the compact byte array as the wrap/update mechanism didn't work as expected, and assumed the smaller format would be preferred. I didn't think the overhead of serializing/deserializing one or the other would be significant. I can migrate back to writing out the updateable byte array, but heapifying the updateable byte array on deserialization.

Reply:
I prefer that we serialize to an updatable byte array. I believe deserializing from that is faster. Thank you for looking into that.